services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/utils/GoalStateManager.java

```diff
@@ -113,221 +213,453 @@ public boolean shouldSkipClass(Class<?> clazz) {
+                    Neighbor.NeighborType l3 = Neighbor.NeighborType.L3;
+                    Neighbor.NeighborType l2 = Neighbor.NeighborType.L2;
+
+                    if (hostIpSubnetIdsMap.get(currentGroupHostIp).size() == 1) {
+                      if (!hostIpSubnetIdsMap
+                          .get(currentGroupHostIp)
+                          .iterator()
+                          .next()
+                          .equals(ipSubnetIdMap.get(ip))) {
+                        l2 = Neighbor.NeighborType.L3;
+                        l3 = Neighbor.NeighborType.L2;
+                      }
+                    }
+
+                    final InternalPortEntity portStateWithEverythingFilledNB =
+                        portIdPortMap.get(ipPortIdMap.get(ip));
+
+                    if (!portIdNeighborInfoMap.containsKey(ip)) {
+                      Set<InternalSubnetEntity> ss =
+                          portStateWithEverythingFilledNB.getSubnetEntities();
+                      final InternalSubnetEntity next = ss.iterator().next();
+                      if (ss.size() == 1 && ipSubnetIdMap.get(ip).equals(next.getId())) {
+
+                        createNeighborState(
+                            networkConfiguration,
+                            neighborStates,
+                            brandNewIps,
+                            ip,
+                            currentGroupHostIp,
+                            l2,
+                            ipPortIdMap,
+                            ipMacMap,
+                            ipSubnetIdMap,
+                            ipHostIpMap,
+                            hostIpFixedIpsMap,
+                            hostIpSubnetIdsMap,
+                            subnetIdSubnetsMap,
+                            portIdPortMap,
+                            portIdNeighborInfoMap);
+                      } else {
+                        createNeighborState(
+                            networkConfiguration,
+                            neighborStates,
+                            brandNewIps,
+                            ip,
+                            currentGroupHostIp,
+                            l3,
+                            ipPortIdMap,
+                            ipMacMap,
+                            ipSubnetIdMap,
+                            ipHostIpMap,
+                            hostIpFixedIpsMap,
+                            hostIpSubnetIdsMap,
+                            subnetIdSubnetsMap,
+                            portIdPortMap,
+                            portIdNeighborInfoMap);
+                      } // end if internal size==1 sn>2
+                      brandNewIps.add(ip);
+                    }
+                  }
+                  // start new loop for adding new ip to existing port
+
+                } // size ==1
+                for (String nip : brandNewIps) {
+                  if (!ipHostIpMap.get(nip).equals(currentGroupHostIp)) continue;
```

Comment:
Is it possible that ipHostIpMap.get(nip) == null?

Reply:
it's impossible since nip <- brandNewIps <- ip <-port which is previously constructed in convert method